pkg/sync/service.go

```diff
@@ -388,6 +388,73 @@ func (s *Service) doDownloadFromURL(request *http.Request) (err error) {
 	return nil
 }
 
+func (s *Service) CloneFromBackingImage(writer http.ResponseWriter, request *http.Request) {
+	err := s.doCloneFromBackingImage(request)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (s *Service) doCloneFromBackingImage(request *http.Request) (err error) {
+	defer func() {
+		if err != nil {
+			s.log.WithError(err).Errorf("Sync Service: failed to do clone from backing image")
+		}
+	}()
+
+	queryParams := request.URL.Query()
+	sourceBackingImage := queryParams.Get(types.DataSourceTypeCloneParameterBackingImage)
+	if sourceBackingImage == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImage)
+	}
+
+	sourceBackingImageUUID := queryParams.Get(types.DataSourceTypeCloneParameterBackingImageUUID)
+	if sourceBackingImageUUID == "" {
+		return fmt.Errorf("%v is not specified", types.DataSourceTypeCloneParameterBackingImageUUID)
+	}
+
+	encryption := types.EncryptionType(queryParams.Get(types.DataSourceTypeCloneParameterEncryption))
+	if encryption != types.EncryptionTypeEncrypt && encryption != types.EncryptionTypeDecrypt && encryption != types.EncryptionTypeIgnore {
+		return fmt.Errorf("%v operation %v is not specified", types.DataSourceTypeCloneParameterEncryption, encryption)
+	}
+	filePath := queryParams.Get("file-path")
+	if filePath == "" {
+		return fmt.Errorf("no filePath restoring file")
+	}
+	uuid := queryParams.Get("uuid")
+	if uuid == "" {
+		return fmt.Errorf("no uuid for restoring file")
+	}
+	diskUUID := queryParams.Get("disk-uuid")
+	expectedChecksum := queryParams.Get("expected-checksum")
+
+	credential := map[string]string{}
+	if err := json.NewDecoder(request.Body).Decode(&credential); err != nil {
+		if err != io.EOF {
+			return fmt.Errorf("get credential failed from request")
+		}
+	}
+
+	sf, err := s.checkAndInitSyncFile(filePath, uuid, diskUUID, expectedChecksum, 0)
+	if err != nil {
+		return err
+	}
+	go func() {
+		if err := sf.WaitForStateNonPending(); err != nil {
+			s.log.Errorf("Sync Service: failed to wait for sync file %v becoming non-pending state before starting the actual cloning: %v", filePath, err)
+			return
+		}
+
+		if _, err := sf.CloneToFileWithEncryption(sourceBackingImage, sourceBackingImageUUID, encryption, credential); err != nil {
+			s.log.Errorf("Sync Service: failed to clone sync file %v: %v", filePath, err)
+			return
```

Comment:
We don't return error here. How is caller or control plane aware of the error?

Reply:
The design of backing image is designed by @shuo-wu quite long time ago
I will briefly introduce it here

1. longhorn-manager create BackingImageDataSource POD/CRD to prepare the backingimage
2. BackingImageDataSource is a service has a function called `Get` to get the backing file status
3. Inside the BackingImageDataSource service there is another service called `sync_file`
    - `sync service` is the one doing the `download`, `upload`, `clone`... etc.
5. sync_file control the state of the backing file and update the status of it
6. longhorn-manager BackingImageDataSource Controller will launch a monitor to get the status of the file
    - monitor -> data source service Get() -> sync_file

So if the file fails in the end, the monitor will know and update the status of the CRD for further processing. 
The return here doesn't matter, this function is called by 
```
router.HandleFunc("/v1/files", service.CloneFromBackingImage).Methods("POST").Queries("action", "cloneFromBackingImage")
```
and it is an async function, it returns 200 back first and starts prepare the backing file, and use the monitor mechanism above to track the status of the backing file.

Reply:
learn more about the backing image mechaninsm
you can check my PPT of tracing the preparation process
https://docs.google.com/presentation/d/1gzzUmkGPb-Ye6J2-5bDoHA89KbjQwD0V0GsaN5u8bRI/edit?usp=sharing